File: apps/cli/internal/db/workspace_types.go
package db

// Workspace is the daemon-owned local workspace record.
type Workspace struct {
	ID             string  `json:"id"`
	OrganizationID string  `json:"organizationId"`
	ProjectID      string  `json:"projectId"`
	NodeID         string  `json:"nodeId"`
	Kind           string  `json:"kind"`
	Status         string  `json:"status"`
	Branch         *string `json:"branch,omitempty"`
	SourceBranch   *string `json:"sourceBranch,omitempty"`
	LocalPath      string  `json:"localPath"`
	State          string  `json:"state"`
	Health         *string `json:"health,omitempty"`
	CreatedAt      string  `json:"createdAt"`
	UpdatedAt      string  `json:"updatedAt"`
}

// WorkspaceUpdate contains the mutable fields of a local workspace.
type WorkspaceUpdate struct {
	Status    *string
	State     *string
	Health    *string
	LocalPath *string
	Branch    *string
}

// WorkspacePullRequest is a pull request observed for a local workspace.
type WorkspacePullRequest struct {
	ID             string  `json:"id"`
	WorkspaceID    string  `json:"workspaceId"`
	OrganizationID string  `json:"organizationId"`
	PRID           string  `json:"prId"`
	Title          *string `json:"title,omitempty"`
	URL            *string `json:"url,omitempty"`
	Branch         *string `json:"branch,omitempty"`
	BaseBranch     *string `json:"baseBranch,omitempty"`
	State          string  `json:"state"`
	Metadata       *string `json:"metadata,omitempty"`
	DetectedAt     string  `json:"detectedAt"`
	ResolvedAt     *string `json:"resolvedAt,omitempty"`
	CreatedAt      string  `json:"createdAt"`
	UpdatedAt      string  `json:"updatedAt"`
}

File: apps/desktop/src/renderer/features/workspace/model/workspaceTypes.ts
/** Feature-owned workspace lifecycle status (replaces api WorkspaceRecord["status"] refs in stores). */
export type WorkspaceStatus = "active" | "closed" | "provisioning";

/**
 * Workspace Store `addWorkspace` action input (desktop6-adjust.md W1).
 * Defined in the Workspace model so Commands, Events, and model builders can
 * use it without importing Store State.
 */
export type AddWorkspaceInput = {
  organizationId?: string;
  projectId?: string;
  repoId?: string;
  name: string;
  sourceBranch: string;
  branch: string;
  worktreePath?: string;
  nodeId?: string;
  workspaceId: string;
  status?: WorkspaceStatus;
  preserveOnMissingSnapshot?: boolean;
};

/**
 * Workspace feature vocabulary (Phase 3 split of features/workbench/model/types.ts).
 * Feature-owned status unions replace transport DTO references.
 */

export type WorkspaceLifecycleState = "active" | "error" | "closing";

export type WorkspaceHealth = "path-missing" | "not-worktree";

export type WorkspaceGitChangeTotals = {
  additions: number;
  deletions: number;
};

export type WorkspaceItem = {
  id: string;
  organizationId?: string;
  projectId?: string;
  repoId: string;
  name: string;
  title: string;
  sourceBranch: string;
  branch: string;
  summaryId: string;
  worktreePath?: string;
  nodeId?: string;
  kind?: "managed" | "local" | "folder";
  status?: WorkspaceStatus;
  preserveOnMissingSnapshot?: boolean;
  state?: WorkspaceLifecycleState;
  health?: WorkspaceHealth;
};

/** Resolves the owning project id for a workspace (folder workspaces use their repo id). */
export function resolveWorkspaceProjectId(workspace: { projectId?: string; repoId: string }): string {
  return workspace.projectId ?? workspace.repoId;
}

/** Resolves the workspace to activate for a project (first workspace of the project, or ""). */
export function resolveWorkspaceIdForProject(workspaces: WorkspaceItem[], projectId: string): string {
  return workspaces.find((workspace) => resolveWorkspaceProjectId(workspace) === projectId)?.id ?? "";
}
